ebi_eva_common_pyutils/assembly/assembly.py

```diff
@@ -0,0 +1,231 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+import urllib
+from csv import DictReader, excel_tab
+from ftplib import FTP
+import re
+from urllib import request
+
+from cached_property import cached_property
+from retry import retry
+
+from ebi_eva_common_pyutils.command_utils import run_command_with_output
+from ebi_eva_common_pyutils.logger import AppLogger
+
+
+class NCBIAssembly(AppLogger):
+    """
+    Class that represent an assembly that would originate from NCBI data
+    It takes a GCA or GCF accession and can download the assembly report and genomics fasta.
+    Using species_scientific_name and assembly_accession it create a directory structure in the provided
+    reference_directory:
+        - species_scientific_name1
+            - assembly_accession1
+            - assembly_accession2
+        - species_scientific_name2
+    the eutils_api_key is only used to retrieve additional contigs if required.
+    """
+
+    def __init__(self, assembly_accession, species_scientific_name, reference_directory, eutils_api_key=None):
+        self.check_assembly_accession_format(assembly_accession)
+        self.assembly_accession = assembly_accession
+        self.species_scientific_name = species_scientific_name
+        self.reference_directory = reference_directory
+        self.eutils_api_key = eutils_api_key
+
+    @staticmethod
+    def check_assembly_accession_format(assembly_accession):
+        if re.match(r"^GC[F|A]_\d+\.\d+$", assembly_accession) is None:
+            raise Exception('Invalid assembly accession: it has to be in the form of '
+                            'GCF_XXXXXXXXX.X or GCA_XXXXXXXXX.X where X is a number')
+
+    @property
+    def assembly_directory(self):
+        assembly_directory = os.path.join(
+            self.reference_directory,  self.species_scientific_name.lower().replace(' ', '_'), self.assembly_accession
+        )
+        os.makedirs(assembly_directory, exist_ok=True),
+        return assembly_directory
+
+    @property
+    def assembly_report_path(self):
+        return os.path.join(self.assembly_directory, self.assembly_accession + '_assembly_report.txt')
+
+    @property
+    def assembly_fasta_path(self):
+        return os.path.join(self.assembly_directory, self.assembly_accession + '.fa')
+
+    @property
+    def assembly_compressed_fasta_path(self):
+        return os.path.join(self.assembly_directory, self.assembly_accession + '.fa.gz')
+
+    @retry(tries=4, delay=2, backoff=1.2, jitter=(1, 3))
+    def _download_file(self, destination_file, url):
+        self.info('Download assembly file for %s to %s', self.assembly_accession, destination_file)
+        request.urlretrieve(url, destination_file)
+        request.urlcleanup()
+
+    @cached_property
+    def _ncbi_genome_folder_url_and_content(self):
```

Comment:
This is more of a judgment call but should this also have the retry decorator?

Reply:
I would have to figure out what Exception I should retrying on. Since We're raising when multiple `genome_subfolders` are found there is a chance it would retry because of that if I catch any exception.
I have not seen any intermittent failure to connect so far (as opposed to eutils) so I'm not too keen.